rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+# RFC 6517 - 2021-03-36 - End-to-end Acknowledgement
+
+This RFC proposes a framework for tracking the flow of events such that the
+source can acknowledge them only after they are successfully delivered.
+
+## Scope
+
+This RFC will cover:
+
+* The functional requirements to support end-to-end acknowledgement
+
+* The data model changes and potential impacts to support the above
+
+
+This RFC will not cover:
+
+* Any details of changes to individual sources, transforms, or sinks to
+  necessary to support this feature.
+
+## Motivation
+
+Users want assurance that their data pipeline successfully delivers all the
+events that are presented to it. To accomplish this, Vector needs to be able to
+maintain information on each event tying it back to the source that originated
+the event, so that source can correctly acknowledge those incoming events at
+the right time.
+
+## Discussion
+
+Vector contains a number of features that can cause complications with a naïve
+implementation. As such, we need to itemize a number of scenarios to ensure our
+proposed solution covers all the edge cases. These will start with the simplest
+scenario and then introduce complicating factors.
+
+In the discussion, the term "finalization" is used to describe the point
+at which an event completes its trip through the pipeline, which may be
+caused by its acknowledged delivery, a permanent failure, or a situation
+causing the event to be dropped in processing.
+
+### Single source through linear transforms to a single sink
+
+This is the simplest of cases. For each event the source produces, the
+event is only modified on its transport through the system. The set of
+transforms represented here are all those not listed explicitly
+below. When the destination sink finishes handling the event, a
```

Comment:
By "finished handling" do you mean the downstream service the sink is interacting with has acked receipt, where the transport supports it? I could also see this being configurable. Maybe I'm happy to have the source message acked once it is persisted to the disk buffer for a given sink.

Reply:
We discussed acking when the event is persisting and I think it was rejected, though I agree that could be useful and likely relatively easy to add.

Reply:
Yes acking when an event reaches the disk buffer is something we want to support.

Reply:
I will note this.